File: atcoder/abc431/e.py
from collections import deque


def bfs01(adj, src, inf):
    n = len(adj)
    dis = [inf for _ in range(n)]
    que = deque()

    dis[src] = 0
    que.append((src, dis[src]))

    while que:
        u, d = que.popleft()
        if d > dis[u]:
            continue
        for v, w in adj[u]:
            nd = d + w
            if dis[v] > nd:
                dis[v] = nd
                if w == 0:
                    que.appendleft((v, nd))
                else:
                    que.append((v, nd))

    return dis


TC = int(input())

#  0
# 1 3
#  2
dirs = "ABC"
valid = [
    [(1, 3), (0, 2), (2, 0), (3, 1)],  # A
    [(1, 2), (0, 3), (2, 1), (3, 0)],  # B
    [(1, 0), (0, 1), (2, 3), (3, 2)],  # C
]

ans = []
for _ in range(TC):
    H, W = map(int, input().split())
    S = [input() for _ in range(H)]

    def id(r, c, d):
        return (r * W + c) * 4 + d

    adj = [[] for _ in range(4 * H * W)]

    # Identity edges
    for r in range(H):
        for c in range(W):
            if r >= 1:
                adj[id(r, c, 0)].append((id(r - 1, c, 2), 0))
            if c >= 1:
                adj[id(r, c, 1)].append((id(r, c - 1, 3), 0))
            if r + 1 < H:
                adj[id(r, c, 2)].append((id(r + 1, c, 0), 0))
            if c + 1 < W:
                adj[id(r, c, 3)].append((id(r, c + 1, 1), 0))

    # Change (r, c) to k
    for r in range(H):
        for c in range(W):
            for k in range(3):
                cost = 0 if S[r][c] == dirs[k] else 1
                for src, dst in valid[k]:
                    adj[id(r, c, src)].append((id(r, c, dst), cost))

    dis = bfs01(adj, id(0, 0, 1), 10**8)
    ans.append(dis[id(H - 1, W - 1, 3)])

print("\n".join(map(str, ans)))

File: atcoder/abc431/e.rs
#![allow(unused)]

use proconio::input;
use proconio::marker::Chars;

fn main() {
    input! {
        tc: usize,
    }

    let dirs = ['A', 'B', 'C'];
    let valid = [
        [(1, 3), (0, 2), (2, 0), (3, 1)], // A
        [(1, 2), (0, 3), (2, 1), (3, 0)], // B
        [(1, 0), (0, 1), (2, 3), (3, 2)], // C
    ];

    let mut ans = vec![];
    for _ in 0..tc {
        input! {
            h: usize,
            w: usize,
            s: [Chars; h],
        }

        //       0
        // 1  (r, c)   3
        //       2
        let id = |r: usize, c: usize, d: usize| (r * w + c) * 4 + d;

        let mut adj = vec![vec![]; h * w * 4];
        for r in 0..h {
            for c in 0..w {
                if r >= 1 {
                    adj[id(r, c, 0)].push((id(r - 1, c, 2), 0));
                }
                if c >= 1 {
                    adj[id(r, c, 1)].push((id(r, c - 1, 3), 0));
                }
                if r + 1 < h {
                    adj[id(r, c, 2)].push((id(r + 1, c, 0), 0));
                }
                if c + 1 < w {
                    adj[id(r, c, 3)].push((id(r, c + 1, 1), 0));
                }
            }
        }

        // change (r, c) to type k
        for r in 0..h {
            for c in 0..w {
                for k in 0..3 {
                    let cost: usize = if s[r][c] == dirs[k] { 0 } else { 1 };
                    for (src, dst) in valid[k] {
                        adj[id(r, c, src)].push((id(r, c, dst), cost));
                    }
                }
            }
        }

        let (dis, _) = bfs01(&adj, id(0, 0, 1), 1_000_000);
        ans.push(dis[id(h - 1, w - 1, 3)]);
    }

    println!("{}", join(&ans, "\n"));
}

use std::collections::VecDeque;

fn bfs01<T>(adj: &Vec<Vec<(usize, T)>>, s: usize, inf: T) -> (Vec<T>, Vec<usize>)
where
    T: std::ops::Add<Output = T> + Copy + Default + Eq + Ord,
{
    let n = adj.len();
    let mut que = VecDeque::new();
    let mut dis = vec![inf; n];
    let mut par = vec![!0; n];

    dis[s] = T::default();
    par[s] = s;
    que.push_back((s, dis[s]));

    while let Some((u, d)) = que.pop_front() {
        if d > dis[u] {
            continue;
        }
        for &(v, w) in &adj[u] {
            let new_d = dis[u] + w;
            if new_d < dis[v] {
                dis[v] = new_d;
                par[v] = u;
                if w == T::default() {
                    que.push_front((v, dis[v]));
                } else {
                    que.push_back((v, dis[v]));
                }
            }
        }
    }

    (dis, par)
}

fn join<T: ToString>(arr: &[T], sep: &str) -> String {
    arr.iter()
        .map(|x| x.to_string())
        .collect::<Vec<String>>()
        .join(sep)
}
